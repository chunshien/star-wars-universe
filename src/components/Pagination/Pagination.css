.pagination {  
    margin-top: 30px;
    text-align: right;  
    -khtml-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    -webkit-touch-callout: none;
    -webkit-user-select: none;                      
}  
.pagination span{
    padding: 5px 10px;
    min-width: 20px;
    cursor: pointer;
}                    
.active-page{
    background-color: white;
    border-bottom: 5px solid black;
}
.enabled-page-navigation{
    color: black;
}
.disabled-page-navigation{
    color: #dddddd;
    cursor: not-allowed !important;
}