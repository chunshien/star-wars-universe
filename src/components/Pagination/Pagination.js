import React, {useCallback} from 'react';
import './Pagination.css';

function Pagination(props) {
    const {activePage, totalItem, itemCountPerPage} = props;
    const lastPage = Math.ceil(totalItem / itemCountPerPage);        
    const firstIndex = activePage > 1 ? 
                (activePage < lastPage ? activePage - 1 : activePage - 2)
                : 1;
    const secondIndex = activePage > 1 ? 
                (activePage < lastPage ? activePage : activePage - 1) 
                : 2;
    const lastIndex = activePage < lastPage ? 
                (activePage > 1 ? activePage + 1 : activePage + 2) 
                : lastPage;
    
    const previousClick = useCallback(() => {
        if(activePage > 1){
            props.onChanged(activePage - 1)
        } 
    }, [props, activePage]);

    const nextClick = useCallback(() => {
        const lastPage = Math.ceil(totalItem / itemCountPerPage);
        if(activePage < lastPage){
            props.onChanged(activePage + 1)
        }   
    }, [props, activePage, totalItem, itemCountPerPage]);

    const pageClick = useCallback((index) => {
        props.onChanged(index)
    }, [props])
    
    return (
        <div className="pagination">
            <span 
                onClick={previousClick} 
                className={activePage === 1 ? 'disabled-page-navigation' : 'enabled-page-navigation'}>
                {'< Previous page'}
            </span>
            <span 
                onClick={()=>pageClick(firstIndex)} 
                className={activePage === 1 ? 'active-page' : ''}>
                {firstIndex}
            </span>
            {Math.ceil(totalItem / itemCountPerPage) >= 2 &&
                <span 
                    onClick={()=>pageClick(secondIndex)} 
                    className={activePage > 1 && activePage < lastPage ? 'active-page' : ''}>
                    {secondIndex}
                </span>
            }
            {Math.ceil(totalItem / itemCountPerPage) >= 3 &&
                <span 
                    onClick={()=>pageClick(lastIndex)} 
                    className={activePage === lastPage ? 'active-page' : ''}>
                    {lastIndex}
                </span>
            }
            <span onClick={nextClick} className={activePage !== lastPage ? 'enabled-page-navigation' : 'disabled-page-navigation'}>
                {'Next page >'}
            </span>
        </div>
    );
    
}
Pagination.defaultProps = {
    activePage: 1,
    totalItem: 0,
    itemCountPerPage: 10
}
export default Pagination;