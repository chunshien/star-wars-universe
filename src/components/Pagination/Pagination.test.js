import React from 'react';
import Pagination from './Pagination.js';
import renderer from 'react-test-renderer';
import TestUtils from 'react-dom/test-utils';

const activePageStyle = {
    backgroundColor: 'white',
    borderBottom: '5px solid black'
};

const enabledPageNavigation = {
    color: 'black'
}

const disabledPageNavigation = {
    color: '#dddddd',
    cursor: 'not-allowed'
}

describe('Component - Pagination Test', () => {
    it('renders Pagination without crashing', () => {
        const component = renderer.create(
            <Pagination />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()        
    });
    
    let PaginationElement
    let onChangedMock = jest.fn();
    class Wrapper extends React.Component {
        render() { 
            return this.props.children
        }
    }
    beforeAll(function() {
        PaginationElement = TestUtils.renderIntoDocument(<Wrapper><Pagination onChanged={onChangedMock}/></Wrapper>);
    });
    beforeEach(function() {
        onChangedMock.mockClear();
    });

    it('Simulate Pagination onChanged', () => {
        PaginationElement.props.children.props.onChanged();
        expect(onChangedMock).toBeCalled();        
    });

    it('Test Pagination 3 pages - page 1 active (first)', () => {
        const component = renderer.create(
            <Pagination 
                totalItem={20}
                itemCountPerPage={8}
                activePage={1}                            
            />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()  
        
        //test previous page
        expect(tree.children[0].children[0]).toEqual('< Previous page')
        expect(tree.children[0].props.style).toEqual(disabledPageNavigation)
        
        //test first index page
        expect(tree.children[1].children[0]).toEqual('1')
        expect(tree.children[1].props.style).toEqual(activePageStyle) // active page
        
        //test second index page        
        expect(tree.children[2].children[0]).toEqual('2')
        expect(tree.children[2].props.style).toEqual({})
        
        //test third index page                
        expect(tree.children[3].children[0]).toEqual('3')
        expect(tree.children[3].props.style).toEqual({})

        //test next page                
        expect(tree.children[4].children[0]).toEqual('Next page >')
        expect(tree.children[4].props.style).toEqual(enabledPageNavigation)
    });

    it('Test Pagination 3 pages - page 2 active (second index)', () => {
        const component = renderer.create(
            <Pagination 
                totalItem={20}
                itemCountPerPage={8}
                activePage={2}                            
            />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()  
        
        //test previous page
        expect(tree.children[0].children[0]).toEqual('< Previous page')
        expect(tree.children[0].props.style).toEqual(enabledPageNavigation)
        
        //test first index page
        expect(tree.children[1].children[0]).toEqual('1')
        expect(tree.children[1].props.style).toEqual({})
        
        //test second index page        
        expect(tree.children[2].children[0]).toEqual('2')
        expect(tree.children[2].props.style).toEqual(activePageStyle) // active page
        
        //test third index page                
        expect(tree.children[3].children[0]).toEqual('3')
        expect(tree.children[3].props.style).toEqual({})

        //test next page                
        expect(tree.children[4].children[0]).toEqual('Next page >')
        expect(tree.children[4].props.style).toEqual(enabledPageNavigation)       
    });

    it('Test Pagination 3 pages - page 3 active (last)', () => {
        const component = renderer.create(
            <Pagination 
                totalItem={20}
                itemCountPerPage={8}
                activePage={3}                            
            />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()  
        
        //test previous page
        expect(tree.children[0].children[0]).toEqual('< Previous page')
        expect(tree.children[0].props.style).toEqual(enabledPageNavigation)
        
        //test first index page
        expect(tree.children[1].children[0]).toEqual('1')
        expect(tree.children[1].props.style).toEqual({})
        
        //test second index page        
        expect(tree.children[2].children[0]).toEqual('2')
        expect(tree.children[2].props.style).toEqual({})
        
        //test third index page                
        expect(tree.children[3].children[0]).toEqual('3')
        expect(tree.children[3].props.style).toEqual(activePageStyle) // active page

        //test next page                
        expect(tree.children[4].children[0]).toEqual('Next page >')
        expect(tree.children[4].props.style).toEqual(disabledPageNavigation)
    });

    it('Test Pagination 4 pages - page 3 active (second index)', () => {
        const component = renderer.create(
            <Pagination 
                totalItem={26}
                itemCountPerPage={8}
                activePage={3}                            
            />    
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot()  
        
        //test previous page
        expect(tree.children[0].children[0]).toEqual('< Previous page')
        expect(tree.children[0].props.style).toEqual(enabledPageNavigation)
        
        //test first index page
        expect(tree.children[1].children[0]).toEqual('2')
        expect(tree.children[1].props.style).toEqual({})
        
        //test second index page        
        expect(tree.children[2].children[0]).toEqual('3')
        expect(tree.children[2].props.style).toEqual(activePageStyle) // active
        
        //test third index page                
        expect(tree.children[3].children[0]).toEqual('4')
        expect(tree.children[3].props.style).toEqual({})

        //test next page                
        expect(tree.children[4].children[0]).toEqual('Next page >')
        expect(tree.children[4].props.style).toEqual(enabledPageNavigation)       
    });
})