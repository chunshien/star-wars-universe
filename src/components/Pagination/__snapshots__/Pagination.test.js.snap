// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - Pagination Test Test Pagination 3 pages - page 1 active (first) 1`] = `
<div
  className="sc-AxjAm pxkaG"
>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "#dddddd",
        "cursor": "not-allowed",
      }
    }
  >
    &lt; Previous page
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "borderBottom": "5px solid black",
      }
    }
  >
    1
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    2
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    3
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    Next page &gt;
  </span>
</div>
`;

exports[`Component - Pagination Test Test Pagination 3 pages - page 2 active (second index) 1`] = `
<div
  className="sc-AxjAm pxkaG"
>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    &lt; Previous page
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    1
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "borderBottom": "5px solid black",
      }
    }
  >
    2
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    3
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    Next page &gt;
  </span>
</div>
`;

exports[`Component - Pagination Test Test Pagination 3 pages - page 3 active (last) 1`] = `
<div
  className="sc-AxjAm pxkaG"
>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    &lt; Previous page
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    1
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    2
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "borderBottom": "5px solid black",
      }
    }
  >
    3
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "#dddddd",
        "cursor": "not-allowed",
      }
    }
  >
    Next page &gt;
  </span>
</div>
`;

exports[`Component - Pagination Test Test Pagination 4 pages - page 3 active (second index) 1`] = `
<div
  className="sc-AxjAm pxkaG"
>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    &lt; Previous page
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    2
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "borderBottom": "5px solid black",
      }
    }
  >
    3
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={Object {}}
  >
    4
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    Next page &gt;
  </span>
</div>
`;

exports[`Component - Pagination Test renders Pagination without crashing 1`] = `
<div
  className="sc-AxjAm pxkaG"
>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "#dddddd",
        "cursor": "not-allowed",
      }
    }
  >
    &lt; Previous page
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "white",
        "borderBottom": "5px solid black",
      }
    }
  >
    1
  </span>
  <span
    className="sc-AxirZ fiXIyU"
    onClick={[Function]}
    style={
      Object {
        "color": "black",
      }
    }
  >
    Next page &gt;
  </span>
</div>
`;
